# Three popelines for genotype imputation of low-coverage whole-genome sequencing.
(1) BaseVar+STITCH: BaseVar for SNP calling, followed by STITCH for genotype imputation; (2) Bcftools+Beagle4: Bcftools for SNP calling, followed by Beagle v4.1 for genotype imputation; (3) GATK+Beagle5: GATK for SNP calling, followed by Beagle v5.1 for genotype imputation.
# Requirements
* basevar
* bcftools v1.10
* python v2.7
* R v3.6.3 requires STITCH package
* java v1.8
* beagle v4 and v5
* GATK v3.7

# Usage
```
./basevar+stitch.sh allbam.list 26
./bcftools+beagle4.sh allbam.list 26
./gatk+beagle5.sh allbam.list 26
```
# Description of argumentsc
* allbam.list 
bam file for genotype imputation including full path

* 26
Imputation on specified chromosomec

# Output file
* time.txt
Time consumed for imputation
* base+stitch.txt/bcf+bg4.txt/gatk+bg5.txt
Consistency and accuracy of imputation generated by different pipelines.

# File Description
* imputeAccuracyID.py
Calculation of imputation accuracy and consistency

* run_STITCH.R
genotype imputation using STITCH
## The folder Example contains the following files:
* allbam.chr60.pass.vcf.gz
Genotype verification using real sequencing vcf files.

* allbam.list
bam file for genotype imputation including full path. The bam file is too large to upload. Please contact the author if necessary.

* id.txt
ID file used for validated imputation accuracy
