# Three popelines for genotype imputation of low-coverage whole-genome sequencing
* (1) BaseVar+STITCH: BaseVar for SNP calling, followed by STITCH for genotype imputation
* (2) Bcftools+Beagle4: Bcftools for SNP calling, followed by Beagle v4.1 for genotype imputation
* (3) GATK+Beagle5: GATK for SNP calling, followed by Beagle v5.1 for genotype imputation
# Requirements
* basevar
* bcftools v1.10
* python v2.7
* R v3.6.3 requires STITCH package
* java v1.8
* beagle v4 and v5
* GATK v3.7

# Usage
```
./basevar+stitch.sh allbam.list 26
./bcftools+beagle4.sh allbam.list 26
./gatk+beagle5.sh allbam.list 26
```
# Description of argumentsc
* allbam.list  
bam file for genotype imputation including full path.
```
/data/songhl/Seq-500/Datadeal/bam/DownDepth/2x/QDH10.marked_duplicates.downsampled.bam
/data/songhl/Seq-500/Datadeal/bam/DownDepth/2x/QDH12.marked_duplicates.downsampled.bam
/data/songhl/Seq-500/Datadeal/bam/DownDepth/2x/QDH13.marked_duplicates.downsampled.bam
/data/songhl/Seq-500/Datadeal/bam/DownDepth/2x/QDH14.marked_duplicates.downsampled.bam
/data/songhl/Seq-500/Datadeal/bam/DownDepth/2x/QDH1526.marked_duplicates.downsampled.bam
/data/songhl/Seq-500/Datadeal/bam/DownDepth/2x/QDH1527.marked_duplicates.downsampled.bam
/data/songhl/Seq-500/Datadeal/bam/DownDepth/2x/QDH1528.marked_duplicates.downsampled.bam
/data/songhl/Seq-500/Datadeal/bam/DownDepth/2x/QDH1529.marked_duplicates.downsampled.bam
/data/songhl/Seq-500/Datadeal/bam/DownDepth/2x/QDH1530.marked_duplicates.downsampled.bam
/data/songhl/Seq-500/Datadeal/bam/DownDepth/2x/QDH1531.marked_duplicates.downsampled.bam
/data/songhl/Seq-500/Datadeal/bam/DownDepth/2x/QDH1532.marked_duplicates.downsampled.bam
...
```
* 26  
Imputation on specified chromosome.

# Output file
* time.txt  
Time consumed for imputation.
```
time: 3h30m49s
```
* base+stitch.txt/bcf+bg4.txt/gatk+bg5.txt  
Consistency and accuracy of imputation generated by different pipelines.
```
ID	Nsnps	Concordance	Correlation
QDH1576	78081	0.872209628463	0.898289678891
QDH1607	74924	0.898070044312	0.921065674104
QDH1627	56748	0.897740889547	0.922503712822
QDH1797	81471	0.885799855163	0.912884800125
QDH1834	56693	0.875716578766	0.905343412999
QDH240	68960	0.889704176334	0.913859673536
QDH6443	79614	0.895118948929	0.918519696953
QDH6474	70131	0.861202606551	0.874181170539
QDH6481	75779	0.896171762626	0.919784803328
QDH6605	76895	0.892476754015	0.914888248437
QDH922	75958	0.851088759578	0.883567856077
2020-10-1567	48131	0.873511873844	0.903959964826
2020-10-1573	49050	0.8899490316	0.914758638534
2020-10-1604	54077	0.897109676942	0.92076108519
2020-10-1608	52790	0.873688198522	0.90231634755
2020-10-1616	52308	0.88785654202	0.912601064499
2020-10-1639	52773	0.887688780247	0.915212187835
2020-10-1646	51922	0.874947035939	0.904086066605
2020-10-319	51144	0.87191068356	0.898974229172
2020-10-340	51641	0.872485040956	0.900989983909
```

# File Description
* imputeAccuracyID.py  
Calculation of imputation accuracy and consistency.

* run_STITCH.R  
Genotype imputation using STITCH.
## The folder Example contains the following files:
* allbam.chr60.pass.vcf.gz  
Genotype verification using real sequencing vcf files.

* allbam.list  
bam file for genotype imputation including full path. The bam file is too large to upload. Please contact the author if necessary.

* id.txt  
ID file used for validated imputation accuracy.
# Publication
Hailiang Song, Tian Dong, Wei Wang, Boyun Jiang, Xiaoyu Yan, Chenfan Geng, Song Bai, Shijian Xu, Hongxia Hu,
Cost-effective genomic prediction of critical economic traits in sturgeons through low-coverage sequencing,
Genomics, 2024, 116 (4): 110874, https://doi.org/10.1016/j.ygeno.2024.110874.
